pkg/networkserver/internal/test/shared/downlink_task_queue.go

```diff
@@ -164,20 +176,28 @@ func handleDownlinkTaskQueueTest(ctx context.Context, q DownlinkTaskQueue) {
 		t.FailNow()
 	}
 
-	expectSlot := func(t *testing.T, expectedID ttnpb.EndDeviceIdentifiers, expectedAt time.Time) {
-		nextPop <- struct{}{}
-
+	// Expect 3 slots
+	receivedSlots := make(map[ttnpb.EndDeviceIdentifiers][]time.Time, 3)
+	for i := 0; i < 3; i++ {
 		t.Helper()
-
 		a := assertions.New(t)
-
 		select {
 		case s := <-slotCh:
-			a.So(s.id, should.Resemble, expectedID)
-			a.So(s.t, should.Equal, expectedAt)
+			receivedSlots[s.id] = append(receivedSlots[s.id], s.t)
 			a.So(s.ctx, should.HaveParentContextOrEqual, popCtx)
 			s.respCh <- TaskPopFuncResponse{}
 
+			select {
+			case err := <-errCh:
+				if !a.So(err, should.BeNil) {
+					t.FailNow()
+				}
+
+			case <-ctx.Done():
+				t.Fatal("Timed out waiting for Pop to return")
+			}
+			nextPop[s.consumerID] <- struct{}{}
```

Comment:
The test failure is caused by this I think.

This will allow the goroutine that `Pop`s to restart the loop, but we don't cancel it at the end. As such, it keeps listening in the background, eating the connections of the client.

Consider having `popCtx` be a cancellable and cleanup the goroutines at the end of the test (similar to the uplink queue test changes).